src/node_url.cc

```diff
@@ -407,29 +407,64 @@ int64_t ParseIPv4Number(const char* start, const char* end) {
   return strtoll(start, nullptr, R);
 }
 
+// https://url.spec.whatwg.org/#ipv4-number-parser
+bool IsIPv4NumberValid(const std::string_view input) {
+  if (input.empty()) {
+    return false;
+  }
+
+  // If a number starts with '0' it might be a number with base 8 or base
+  // 16. If not, checking if all characters are digits proves that it is a
+  // base 10 number.
+  if (input.size() >= 2 && input[0] == '0') {
+    if (input[1] == 'X' || input[1] == 'x') {
+      if (input.size() == 2) {
+        return true;
+      }
+
+      return input.find_first_not_of("0123456789abcdefABCDEF", 2) ==
+             std::string_view::npos;
+    }
+
+    return input.find_first_not_of("01234567", 1) == std::string_view::npos;
+  }
+
+  return std::all_of(input.begin(), input.end(), ::isdigit);
+}
+
 // https://url.spec.whatwg.org/#ends-in-a-number-checker
-bool EndsInANumber(const std::string& input) {
-  std::vector<std::string> parts = SplitString(input, '.', false);
+inline bool EndsInANumber(const std::string_view input) {
+  if (input.empty()) {
+    return false;
+  }
+
+  char delimiter = '.';
+  std::string_view::iterator pointer_start = input.begin();
+  std::string_view::iterator pointer_end = input.end();
 
-  if (parts.empty()) return false;
+  uint8_t parts_size = std::count(pointer_start, pointer_end, delimiter);
+  ++parts_size;
 
-  if (parts.back().empty()) {
-    if (parts.size() == 1) return false;
-    parts.pop_back();
+  if (input.back() == delimiter) {
+    --pointer_end;
+    --parts_size;
   }
 
-  const std::string& last = parts.back();
+  if (parts_size > 1) {
+    pointer_start +=
+        input.rfind(delimiter, pointer_end - pointer_start - 1) + 1;
+  }
 
-  // If last is non-empty and contains only ASCII digits, then return true
-  if (!last.empty() && std::all_of(last.begin(), last.end(), ::isdigit)) {
-    return true;
+  if ((pointer_end - pointer_start) == 0) {
+    return false;
   }
 
-  const char* last_str = last.c_str();
-  int64_t num = ParseIPv4Number(last_str, last_str + last.size());
-  if (num >= 0) return true;
+  if (std::all_of(pointer_start, pointer_end, ::isdigit)) {
+    return true;
+  }
 
-  return false;
+  return IsIPv4NumberValid(
+      std::string_view(pointer_start, pointer_end - pointer_start));
```

Comment:
Can you cast the second parameter to `size_t`?

```suggestion
      std::string_view(pointer_start, size_t(pointer_end - pointer_start)));
```